core/c4diagram.go

```diff
@@ -0,0 +1,355 @@
+func NewPlantUMLClient(optFns ...func(*optionsPlantUMLClient)) ClientGraphToDiagram {
+	o := optionsPlantUMLClient{
+		httpClient: nil,
+	}
+
+	for _, fn := range optFns {
+		fn(&o)
+	}
+
+	resolveHTTPClient(&o)
+
+	return &clientPlantUML{
+		options: o,
+		baseURL: baseURLPlanUML,
+	}
+}
+
+// WithHTTPClientPlantUML overwrite the PlantUML HTTP client.
+func WithHTTPClientPlantUML(c HttpClient) func(*optionsPlantUMLClient) {
+	return func(o *optionsPlantUMLClient) {
+		o.httpClient = c
+	}
+}
+
+func resolveHTTPClient(o *optionsPlantUMLClient) {
+	if o.httpClient == nil {
+		o.httpClient = &http.Client{Timeout: defaultTimeoutPlanUML}
+	}
+}
+
+// code2Path converts the diagram as code to the 64Bytes encoded string to query plantuml
+//
+// Example: the diagram's code
+// @startuml
+//
+//	a -> b
+//
+// @enduml
+//
+// will be converted to SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+//
+// The resulting string to be used to generate C4 diagram
+// - as png: GET www.plantuml.com/plantuml/png/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+// - as svg: GET www.plantuml.com/plantuml/svg/SoWkIImgAStDuL80WaG5NJk592w7rBmKe100
+func code2Path(s string) (string, error) {
+	zb, err := compress(*(*[]byte)(unsafe.Pointer(&s)))
+	if err != nil {
+		return "", err
+	}
+	return encode64(zb), nil
+}
+
+func compress(v []byte) ([]byte, error) {
+	var options = compression.DefaultOptions()
+	var w bytes.Buffer
+	if err := compression.Compress(&options, compression.FORMAT_DEFLATE, v, &w); err != nil {
+		return nil, err
+	}
+	return w.Bytes(), nil
+}
+
+func encode64(e []byte) string {
+	var r bytes.Buffer
+	for i := 0; i < len(e); i += 3 {
+		switch len(e) {
+		case i + 2:
+			r.Write(append3bytes(e[i], e[i+1], 0))
+		case i + 1:
+			r.Write(append3bytes(e[i], 0, 0))
+		default:
+			r.Write(append3bytes(e[i], e[i+1], e[i+2]))
+		}
+	}
+	return r.String()
+}
+
+func append3bytes(e, n, t byte) []byte {
+	c1 := e >> 2
+	c2 := (3&e)<<4 | n>>4
+	c3 := (15&n)<<2 | t>>6
+	c4 := 63 & t
+
+	var buf bytes.Buffer
+
+	buf.WriteByte(encode6bit(c1 & 63))
+	buf.WriteByte(encode6bit(c2 & 63))
+	buf.WriteByte(encode6bit(c3 & 63))
+	buf.WriteByte(encode6bit(c4 & 63))
+
+	return buf.Bytes()
+}
+
+func encode6bit(e byte) byte {
+	if e < 10 {
+		return 48 + e
+	}
+
+	e -= 10
+	if e < 26 {
+		return 65 + e
+	}
+
+	e -= 26
+	if e < 26 {
+		return 97 + e
+	}
+
+	e -= 26
+	switch e {
+	case 0:
+		return '-'
+	case 1:
+		return '_'
+	default:
+		return '?'
+	}
+}
+
+func trimQuotes(s string) string {
+	if s[:1] == `"` {
+		s = s[1:]
+	}
+	if s[len(s)-1:] == `"` {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+func diagramFooter2UML(graph DiagramGraph) string {
+	if graph.Footer == "" {
+		return `footer "generated by diagramastext.dev - %date('yyyy-MM-dd')"`
+	}
+	return `footer "` + stringCleaner(graph.Footer) + `"`
+}
+
+func diagramNode2UML(n *Node) (string, error) {
+	if n.ID == "" {
+		return "", errors.New("node must be identified: 'id' attribute")
+	}
+
+	o := "Container"
```

Comment:
**suggestion** use `strings.Builder` to create this string efficiently.

Reply:
For refactoring

Reply:
Added "FIXME" for later refactoring.